kategory-core/src/main/kotlin/kategory/data/SortedMapKW.kt

```diff
@@ -0,0 +1,77 @@
+package kategory
+
+import java.util.*
+
+@higherkind
+data class SortedMapKW<K: Comparable<K>, A>(val map: SortedMap<K, A>) : SortedMapKWKind<K, A>, SortedMap<K, A> by map {
+
+    fun <B> map(f: (A) -> B): SortedMapKW<K, B> =
+            this.map.map { it.key to f(it.value) }.toMap().toSortedMap().k()
+
+    fun <B, Z> map2(fb: SortedMapKW<K, B>, f: (A, B) -> Z): SortedMapKW<K, Z> =
+            if (fb.isEmpty()) sortedMapOf<K, Z>().k()
+            else this.map.flatMap { (k, a) ->
+                fb.getOption(k).map { Tuple2(k, f(a, it)) }.k().asIterable()
+            }.k()
+
+    fun <B, Z> map2Eval(fb: Eval<SortedMapKW<K, B>>, f: (A, B) -> Z): Eval<SortedMapKW<K, Z>> =
+            if (fb.isEmpty()) Eval.now(sortedMapOf<K, Z>().k())
+            else fb.map { b -> this.map2(b, f) }
+
+    fun <B> ap(ff: SortedMapKW<K, (A) -> B>): SortedMapKW<K, B> =
+            ff.flatMap { this.map(it) }
+
+    fun <B, Z> ap2(f: SortedMapKW<K, (A, B) -> Z>, fb: SortedMapKW<K, B>): SortedMap<K, Z> =
+            f.map.flatMap { (k, f) ->
+                this.flatMap { a -> fb.flatMap { b -> sortedMapOf(k to f(a, b)).k() } }
+                        .getOption(k).map { Tuple2(k, it) }.k().asIterable()
+            }.k()
+
+    fun <B> flatMap(f: (A) -> SortedMapKW<K, B>): SortedMapKW<K, B> =
+            this.map.flatMap { (k, v) ->
+                f(v).getOption(k).map { Tuple2(k, it) }.k().asIterable()
+            }.k()
+
+    fun <B> foldR(b: Eval<B>, f: (A, Eval<B>) -> Eval<B>): Eval<B> =
+            this.map.values.iterator().iterateRight(b)(f)
+
+    fun <B> foldL(b: B, f: (B, A) -> B): B = this.map.values.fold(b, f)
+
+    fun <B> foldLeft(b: SortedMapKW<K, B>, f: (SortedMapKW<K, B>, Tuple2<K, A>) -> SortedMapKW<K, B>): SortedMapKW<K, B> =
+            this.map.foldLeft(b) { m: SortedMap<K, B>, (k, v) -> f(m.k(), Tuple2(k, v)) }.k()
+
+    fun <G, B> traverse(f: (A) -> HK<G, B>, GA: Applicative<G>): HK<G, SortedMapKW<K, B>> =
+        Foldable.iterateRight(this.map.iterator(), Eval.always { GA.pure(sortedMapOf<K, B>().k()) })({
+            kv, lbuf ->
+            GA.map2Eval(f(kv.value), lbuf) { (sortedMapOf(kv.key to it.a).k() + it.b).toSortedMap().k() }
```

Comment:
If we're going to sort it afterwards, is there any perf gain from using `(mapOf(kv.key to it.a).k() + it.b)` here instead?

Reply:
I don't really know, but there's no need for the first map to be sorted if we are sorting it right after as you say. I'm changing this 👍